File: json/json-annotation-processor/src/test/java/ru/tinkoff/kora/json/annotation/processor/JsonNullableWriteTests.java
package ru.tinkoff.kora.json.annotation.processor;

import com.fasterxml.jackson.core.JsonGenerator;
import org.jetbrains.annotations.Nullable;
import org.junit.jupiter.api.Test;
import ru.tinkoff.kora.json.common.JsonNullable;
import ru.tinkoff.kora.json.common.JsonWriter;
import ru.tinkoff.kora.json.common.ListJsonWriter;

import java.io.IOException;
import java.sql.Timestamp;
import java.time.Instant;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class JsonNullableWriteTests extends AbstractJsonAnnotationProcessorTest {

    @Test
    public void jsonWriterNativeNullableIsUndefined() {
        compile("""
            @JsonWriter
            public record TestRecord(@JsonField("test_field") JsonNullable<String> testField){}
            """);

        var o = writer("TestRecord").toStringUnchecked(newObject("TestRecord", JsonNullable.undefined()));

        assertThat(o).isEqualTo("""
            {}""");
    }

    @Test
    public void jsonWriterNativeNullableIsNullable() {
        compile("""
            @JsonWriter
            public record TestRecord(@JsonField("test_field") JsonNullable<String> testField){}
            """);

        var o = writer("TestRecord").toStringUnchecked(newObject("TestRecord", JsonNullable.nullValue()));

        assertThat(o).isEqualTo("""
            {"test_field":null}""");
    }

    @Test
    public void jsonWriterNativeNullableIsPresent() {
        compile("""
            @JsonWriter
            public record TestRecord(@JsonField("test_field") JsonNullable<String> testField){}
            """);

        var o = writer("TestRecord").toStringUnchecked(newObject("TestRecord", JsonNullable.of("test")));

        assertThat(o).isEqualTo("""
            {"test_field":"test"}""");
    }

    @Test
    public void jsonWriterUserNullableIsUndefined() {
        compile("""
            @JsonWriter
            public record TestRecord(@JsonField("test_field") JsonNullable<java.sql.Timestamp> testField){}
            """);

        var timeWriter = new JsonWriter<Timestamp>() {
            @Override
            public void write(JsonGenerator generator, @Nullable Timestamp object) throws IOException {
                if (object != null) {
                    generator.writeNumber(object.getTime());
                }
            }
        };

        var o = writer("TestRecord", timeWriter).toStringUnchecked(newObject("TestRecord", JsonNullable.undefined()));

        assertThat(o).isEqualTo("""
            {}""");
    }

    @Test
    public void jsonWriterUserNullableIsNullable() {
        compile("""
            @JsonWriter
            public record TestRecord(@JsonField("test_field") JsonNullable<java.sql.Timestamp> testField){}
            """);

        var timeWriter = new JsonWriter<Timestamp>() {
            @Override
            public void write(JsonGenerator generator, @Nullable Timestamp object) throws IOException {
                if (object != null) {
                    generator.writeNumber(object.getTime());
                }
            }
        };

        var o = writer("TestRecord", timeWriter).toStringUnchecked(newObject("TestRecord", JsonNullable.nullValue()));

        assertThat(o).isEqualTo("""
            {"test_field":null}""");
    }

    @Test
    public void jsonWriterUserNullableIsPresent() {
        compile("""
            @JsonWriter
            public record TestRecord(@JsonField("test_field") JsonNullable<java.sql.Timestamp> testField){}
            """);

        var timeWriter = new JsonWriter<Timestamp>() {
            @Override
            public void write(JsonGenerator generator, @Nullable Timestamp object) throws IOException {
                if (object != null) {
                    generator.writeNumber(object.getTime());
                }
            }
        };

        var o = writer("TestRecord", timeWriter).toStringUnchecked(newObject("TestRecord", JsonNullable.of(Timestamp.from(Instant.ofEpochMilli(1)))));

        assertThat(o).isEqualTo("""
            {"test_field":1}""");
    }

    @Test
    public void jsonWriterUserNullableNotEmptyIsEmpty() {
        compile("""
            @JsonWriter
            public record TestRecord(@JsonInclude(IncludeType.NON_EMPTY) @JsonField("test_field") JsonNullable<java.util.List<java.sql.Timestamp>> testField){}
            """);

        var timeWriter = new ListJsonWriter<>((JsonWriter<Timestamp>) (generator, object) -> {
            if (object != null) {
                generator.writeNumber(object.getTime());
            }
        });

        var o = writer("TestRecord", timeWriter).toStringUnchecked(newObject("TestRecord", JsonNullable.of(List.of())));

        assertThat(o).isEqualTo("""
            {}""");
    }

    @Test
    public void jsonWriterUserNullableNotEmptyIsPresent() {
        compile("""
            @JsonWriter
            public record TestRecord(@JsonInclude(IncludeType.NON_EMPTY) @JsonField("test_field") JsonNullable<java.util.List<java.sql.Timestamp>> testField){}
            """);

        var timeWriter = new ListJsonWriter<>((JsonWriter<Timestamp>) (generator, object) -> {
            if (object != null) {
                generator.writeNumber(object.getTime());
            }
        });

        var o = writer("TestRecord", timeWriter).toStringUnchecked(newObject("TestRecord", JsonNullable.of(List.of(Timestamp.from(Instant.ofEpochMilli(1))))));

        assertThat(o).isEqualTo("""
            {"test_field":[1]}""");
    }

    @Test
    public void jsonWriterUserNullableAlwaysIsEmpty() {
        compile("""
            @JsonWriter
            public record TestRecord(@JsonInclude(IncludeType.ALWAYS) @JsonField("test_field") JsonNullable<java.util.List<java.sql.Timestamp>> testField){}
            """);

        var timeWriter = new ListJsonWriter<>((JsonWriter<Timestamp>) (generator, object) -> {
            if (object != null) {
                generator.writeNumber(object.getTime());
            }
        });

        var o = writer("TestRecord", timeWriter).toStringUnchecked(newObject("TestRecord", JsonNullable.of(List.of())));

        assertThat(o).isEqualTo("""
            {"test_field":[]}""");
    }

    @Test
    public void jsonWriterUserNullableAlwaysIsPresent() {
        compile("""
            @JsonWriter
            public record TestRecord(@JsonInclude(IncludeType.ALWAYS) @JsonField("test_field") JsonNullable<java.util.List<java.sql.Timestamp>> testField){}
            """);

        var timeWriter = new ListJsonWriter<>((JsonWriter<Timestamp>) (generator, object) -> {
            if (object != null) {
                generator.writeNumber(object.getTime());
            }
        });

        var o = writer("TestRecord", timeWriter).toStringUnchecked(newObject("TestRecord", JsonNullable.of(List.of(Timestamp.from(Instant.ofEpochMilli(1))))));

        assertThat(o).isEqualTo("""
            {"test_field":[1]}""");
    }

}

File: json/json-symbol-processor/src/test/kotlin/ru/tinkoff/kora/json/ksp/JsonNullableWriteTests.kt
package ru.tinkoff.kora.json.ksp

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import ru.tinkoff.kora.json.common.JsonNullable
import ru.tinkoff.kora.json.common.JsonValue
import ru.tinkoff.kora.json.common.JsonWriter
import ru.tinkoff.kora.json.common.ListJsonWriter
import tools.jackson.core.JsonGenerator
import java.sql.Timestamp
import java.time.Instant

class JsonNullableWriteTests : AbstractJsonSymbolProcessorTest() {

    @Test
    fun jsonWriterNativeNullableIsUndefined() {
        compile(
            """
            @JsonWriter
            data class TestRecord(@field:JsonField("test_field") val testField: JsonNullable<String>)
            """.trimIndent()
        )

        val o = writer("TestRecord").toString(new("TestRecord", JsonValue.undefined<Any>()))

        assertThat(o).isEqualTo(
            """
            {}
            """.trimIndent()
        )
    }

    @Test
    fun jsonWriterNativeNullableIsNullable() {
        compile(
            """
            @JsonWriter
            data class TestRecord(@field:JsonField("test_field") val testField: JsonNullable<String>)
            """.trimIndent()
        )

        val o = writer("TestRecord").toString(new("TestRecord", JsonNullable.nullValue<Any>()))

        assertThat(o).isEqualTo(
            """
            {"test_field":null}
            """.trimIndent()
        )
    }

    @Test
    fun jsonWriterNativeNullableIsPresent() {
        compile(
            """
            @JsonWriter
            data class TestRecord(@field:JsonField("test_field") val testField: JsonNullable<String>)
            """.trimIndent()
        )

        val o = writer("TestRecord").toString(new("TestRecord", JsonNullable.of("test")))

        assertThat(o).isEqualTo(
            """
            {"test_field":"test"}
            """.trimIndent()
        )
    }

    @Test
    fun jsonWriterUserNullableIsUndefined() {
        compile(
            """
            @JsonWriter
            data class TestRecord(@field:JsonField("test_field") val testField: JsonNullable<java.sql.Timestamp>)
            """.trimIndent()
        )

        val timeWriter: JsonWriter<Timestamp> = JsonWriter { generator, `object` ->
            if (`object` != null) {
                generator.writeNumber(`object`.time)
            }
        }

        val o = writer("TestRecord", timeWriter).toString(new("TestRecord", JsonValue.undefined<Any>()))

        assertThat(o).isEqualTo(
            """
            {}
            """.trimIndent()
        )
    }

    @Test
    fun jsonWriterUserNullableIsNullable() {
        compile(
            """
            @JsonWriter
            data class TestRecord(@field:JsonField("test_field") val testField: JsonNullable<java.sql.Timestamp>)
            """.trimIndent()
        )

        val timeWriter: JsonWriter<Timestamp> = JsonWriter { generator, `object` ->
            if (`object` != null) {
                generator.writeNumber(`object`.time)
            }
        }

        val o = writer("TestRecord", timeWriter).toString(new("TestRecord", JsonNullable.nullValue<Any>()))

        assertThat(o).isEqualTo(
            """
            {"test_field":null}
            """.trimIndent()
        )
    }

    @Test
    fun jsonWriterUserNullableIsPresent() {
        compile(
            """
            @JsonWriter
            data class TestRecord(@field:JsonField("test_field") val testField: JsonNullable<java.sql.Timestamp>)
            """.trimIndent()
        )

        val timeWriter: JsonWriter<Timestamp> = JsonWriter { generator, `object` ->
            if (`object` != null) {
                generator.writeNumber(`object`.time)
            }
        }

        val o = writer("TestRecord", timeWriter).toString(new("TestRecord", JsonNullable.of(Timestamp.from(Instant.ofEpochMilli(1)))))

        assertThat(o).isEqualTo(
            """
            {"test_field":1}
            """.trimIndent()
        )
    }

    @Test
    fun jsonWriterUserInnerNullableIsUndefined() {
        compile(
            """
            @JsonWriter
            data class TestRecord(@field:JsonField("test_field") val testField: JsonNullable<java.sql.Timestamp>, val inner: JsonNullable<InnerRecord>)
            """.trimIndent(),
            """
            @JsonWriter
            data class InnerRecord(val simpleField: JsonNullable<java.sql.Timestamp>) 
            """.trimIndent(),
        )

        val timeWriter: JsonWriter<Timestamp> = JsonWriter { generator, `object` ->
            if (`object` != null) {
                generator.writeNumber(`object`.time)
            }
        }

        val o = writer("TestRecord", timeWriter, writer("InnerRecord", timeWriter)).toString(
            new("TestRecord", JsonValue.undefined<Any>(), JsonValue.undefined<Any>()))

        assertThat(o).isEqualTo(
            """
            {}
            """.trimIndent()
        )
    }

    @Test
    fun jsonWriterUserInnerNullableIsNullable() {
        compile(
            """
            @JsonWriter
            data class TestRecord(@field:JsonField("test_field") val testField: JsonNullable<java.sql.Timestamp>, val inner: JsonNullable<InnerRecord>)
            """.trimIndent(),
            """
            @JsonWriter
            data class InnerRecord(val simpleField: JsonNullable<java.sql.Timestamp>) 
            """.trimIndent(),
        )

        val timeWriter: JsonWriter<Timestamp> = JsonWriter { generator, `object` ->
            if (`object` != null) {
                generator.writeNumber(`object`.time)
            }
        }

        val o = writer("TestRecord", timeWriter, writer("InnerRecord", timeWriter)).toString(
            new("TestRecord", JsonNullable.nullValue<Any>(), JsonNullable.nullValue<Any>()))

        assertThat(o).isEqualTo(
            """
            {"test_field":null,"inner":null}
            """.trimIndent()
        )
    }

    @Test
    fun jsonWriterUserInnerNullableIsPresent() {
        compile(
            """
            @JsonWriter
            data class TestRecord(@field:JsonField("test_field") val testField: JsonNullable<java.sql.Timestamp>, val inner: JsonNullable<InnerRecord>)
            """.trimIndent(),
            """
            @JsonWriter
            data class InnerRecord(val simpleField: JsonNullable<java.sql.Timestamp>) 
            """.trimIndent(),
            )

        val timeWriter: JsonWriter<Timestamp> = JsonWriter { generator, `object` ->
            if (`object` != null) {
                generator.writeNumber(`object`.time)
            }
        }

        val o = writer("TestRecord", timeWriter, writer("InnerRecord", timeWriter)).toString(
            new("TestRecord", JsonNullable.of(Timestamp.from(Instant.ofEpochMilli(1))),
                JsonNullable.of(new("InnerRecord", JsonNullable.of(Timestamp.from(Instant.ofEpochMilli(1)))))))

        assertThat(o).isEqualTo(
            """
            {"test_field":1,"inner":{"simpleField":1}}
            """.trimIndent()
        )
    }

    @Test
    fun jsonWriterUserNullableNotEmptyIsEmpty() {
        compile(
            """
            @JsonWriter
            data class TestRecord(@field:JsonInclude(IncludeType.NON_EMPTY) @field:JsonField("test_field") val testField: JsonNullable<List<java.sql.Timestamp>>)
            """.trimIndent()
        )

        val timeWriter = ListJsonWriter { generator: JsonGenerator, `object`: Timestamp? ->
            if (`object` != null) {
                generator.writeNumber(`object`.time)
            }
        }

        val o = writer("TestRecord", timeWriter).toString(new("TestRecord", JsonNullable.of<List<Any>>(listOf())))

        assertThat(o).isEqualTo(
            """
            {}
            """.trimIndent()
        )
    }

    @Test
    fun jsonWriterUserNullableNotEmptyIsPresent() {
        compile(
            """
            @JsonWriter
            data class TestRecord(@field:JsonInclude(IncludeType.NON_EMPTY) @field:JsonField("test_field") val testField: JsonNullable<List<java.sql.Timestamp>>)
            """.trimIndent()
        )

        val timeWriter = ListJsonWriter { generator: JsonGenerator, `object`: Timestamp? ->
            if (`object` != null) {
                generator.writeNumber(`object`.time)
            }
        }

        val o = writer("TestRecord", timeWriter).toString(new("TestRecord", JsonNullable.of<List<Timestamp>>(listOf(Timestamp.from(Instant.ofEpochMilli(1))))))

        assertThat(o).isEqualTo(
            """
            {"test_field":[1]}
            """.trimIndent()
        )
    }

    @Test
    fun jsonWriterUserNullableAlwaysIsEmpty() {
        compile(
            """
            @JsonWriter
            data class TestRecord(@field:JsonInclude(IncludeType.ALWAYS) @field:JsonField("test_field") val testField: JsonNullable<List<java.sql.Timestamp>>)
            """.trimIndent()
        )

        val timeWriter = ListJsonWriter { generator: JsonGenerator, `object`: Timestamp? ->
            if (`object` != null) {
                generator.writeNumber(`object`.time)
            }
        }

        val o = writer("TestRecord", timeWriter).toString(new("TestRecord", JsonNullable.of<List<Any>>(listOf())))

        assertThat(o).isEqualTo(
            """
            {"test_field":[]}
            """.trimIndent()
        )
    }

    @Test
    fun jsonWriterUserNullableAlwaysIsPresent() {
        compile(
            """
            @JsonWriter
            data class TestRecord(@field:JsonInclude(IncludeType.ALWAYS) @field:JsonField("test_field") val testField: JsonNullable<List<java.sql.Timestamp>>)
            """.trimIndent()
        )

        val timeWriter = ListJsonWriter { generator: JsonGenerator, `object`: Timestamp? ->
            if (`object` != null) {
                generator.writeNumber(`object`.time)
            }
        }

        val o = writer("TestRecord", timeWriter).toString(new("TestRecord", JsonNullable.of<List<Timestamp>>(listOf(Timestamp.from(Instant.ofEpochMilli(1))))))

        assertThat(o).isEqualTo(
            """
            {"test_field":[1]}
            """.trimIndent()
        )
    }
}
